### DISCUSSION TITLE
`persist` triggers insert statements before `flush` in some cases

### DISCUSSION THREAD

**Discussion by maphe** [mikro-orm/mikro-orm#2861] [Q&A]
**Describe the bug**
I'm facing a case where the insert statements are executed at persist, and before my code reaches the flush statement.

**To Reproduce**
For context, I'm working on a migration aiming to move some data from a table to another. Here's a simplified version:
```ts
for (const source of sources) {
  const migratedData = { name: source.name, email: source.email };

  const existing = await destRepo.findOne(migratedData);
  if (existing) continue; // so we skip this row if it's already been migrated

  const destObj = destRepo.create(migratedData);
  destRepo.persist(); // <----- inserts happen here
}

await destRepo.flush(); // <----- nothing happens here
```

**Expected behavior**
The insert statements should not occur until we're out of the loop, and do them all at once.

**Additional context**
It works fine when I remove the check for `existing`. So I'm guessing the EM considers the object matching these criteria "managed", even though it returns null (since we skip the `continue` statement). Then for some reason persisting a new object with matching criteria triggers an insert.

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | 4.3.5 |
| mikro-orm | 5.0.5 |
| your-driver | MySQL |

**Comment by B4nan** [comment]
That is a feature, not a bug. Its not the persist call, sync calls cant load anything - its the findOne. 

https://mikro-orm.io/docs/unit-of-work#flush-modes

**Reply by maphe** [reply]
Ok, thanks for the answer. I did not realize it inserts the object from the previous loop iteration.

The doc mentions
> `FlushMode.AUTO` - This is the default mode, and it flushes the `EntityManager` only if necessary.

Out of curiosity, what makes it consider it's necessary in this case?

Also, when I try `em.fork({ flushMode: FlushMode.COMMIT })` I get
```
Cannot read properties of undefined (reading 'COMMIT')
```
I had to do `em.fork({ flushMode: 0 })` for it to work. Not sure if there's an export issue on your side.

**Reply by B4nan** [reply]
The heuristic is rather simple, if there is any dirty entity of such type, it will trigger flushing. This means either persist call of new entity or a change of state of loaded entity. 

The issue with undefined should be fixed in latest dev version, its because its defined as const enum.

**Reply by maphe** [reply]
Got it, thanks. That's a new concept with v5 right? I never faced that before migrating to v5, although I'm sure we do have some code that has similar logic.
